File: JS/LeetC/g75/lowestCommonAncestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */


// see iterative approach in python version
function lowestCommonAncestor(root, p, q) {
  if (!root) {
    return root;
  }

  if (p.val < root.val && q.val < root.val) {
    return lowestCommonAncestor(root.left, p, q);
  }

  if (p.val > root.val && q.val > root.val) {
    return lowestCommonAncestor(root.right, p, q);
  }

  return root;
}

File: JS/LeetC/g75/lowestCommonAncestor.py
class Solution:

    def lowestCommonAncestor(self, root, p, q):

        p_val = p.val
        q_val = q.val

        node = root

        while node:

            parent_val = node.val

            if p_val > parent_val and q_val > parent_val:
                node = node.right
            elif p_val < parent_val and q_val < parent_val:
                node = node.left
            else:
                return node
